bucket_map/src/restart.rs

```diff
@@ -0,0 +1,262 @@
+        let header = self.get_header();
+        writeln!(f, "{:?}", header)?;
+        write!(
+            f,
+            "{:?}",
+            (0..header.buckets)
+                .map(|index| self.get_bucket(index))
+                .take(10)
+                .collect::<Vec<_>>()
+        )?;
+        Ok(())
+    }
+}
+
+impl Restart {
+    /// create a new restart file for use next time we restart on this machine
+    pub(crate) fn new(config: &BucketMapConfig) -> Option<Restart> {
+        let expected_len = Self::expected_len(config.max_buckets);
+
+        let path = config.restart_config_file.as_ref();
+        let path = path?;
+        _ = remove_file(path);
+
+        let mmap = Self::new_map(path, expected_len as u64).ok()?;
+
+        let mut restart = Restart { mmap };
+        let header = restart.get_header_mut();
+        header.version = HEADER_VERSION;
+        header.buckets = config.max_buckets;
+        header.max_search = config.max_search.unwrap_or(MAX_SEARCH_DEFAULT);
+
+        (0..config.max_buckets).for_each(|index| {
+            let bucket = restart.get_bucket_mut(index);
+            bucket.file_name = 0;
+            bucket.random = 0;
+        });
+
+        Some(restart)
+    }
+
+    /// expected len of file given this many buckets
+    fn expected_len(max_buckets: usize) -> usize {
+        std::mem::size_of::<Header>() + max_buckets * std::mem::size_of::<OneIndexBucket>()
+    }
+
+    /// create mmap from `file`
+    fn new_map(file: impl AsRef<Path>, capacity: u64) -> Result<MmapMut, std::io::Error> {
+        let mut data = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .create(true)
+            .open(file)?;
+
+        // Theoretical performance optimization: write a zero to the end of
+        // the file so that we won't have to resize it later, which may be
+        // expensive.
+        data.seek(SeekFrom::Start(capacity - 1)).unwrap();
+        data.write_all(&[0]).unwrap();
+        data.rewind().unwrap();
+        data.flush().unwrap();
+        Ok(unsafe { MmapMut::map_mut(&data).unwrap() })
+    }
+
+    fn get_header(&self) -> &Header {
+        let start = 0_usize;
+        let end = start + std::mem::size_of::<Header>();
+        let item_slice: &[u8] = &self.mmap[start..end];
+        unsafe {
+            let item = item_slice.as_ptr() as *const Header;
+            &*item
+        }
+    }
+
+    fn get_header_mut(&mut self) -> &mut Header {
+        let start = 0_usize;
+        let end = start + std::mem::size_of::<Header>();
+        let item_slice: &[u8] = &self.mmap[start..end];
+        unsafe {
+            let item = item_slice.as_ptr() as *mut Header;
+            &mut *item
+        }
+    }
+
+    fn get_bucket(&self, index: usize) -> &OneIndexBucket {
+        let record_len = std::mem::size_of::<OneIndexBucket>();
+        let start = std::mem::size_of::<Header>() + record_len * index;
```

Comment:
nit: how about assert that index < header.buckects?

Reply:
we assert accessing the slice if `index` is too big. Don't ask me how I know ;-)

Reply:
ok.